File: QUIckControl/QUIckControl/QUIckControl.m
//
//  KDNControl.m
//  KDNControl
//
//  Created by Denis Koryttsev on 16/10/16.
//  Copyright © 2016 Denis Koryttsev. All rights reserved.
//

#import "QUIckControl.h"
#import "QUIckControlPrivate.h"

@interface QUIckControl ()
@property (nonatomic) BOOL isTransitionTime;
@property (nonatomic, strong) QUIckControlValueTarget * thisValueTarget;
@property (nonatomic, strong) NSMutableArray * states;
@property (nonatomic, strong) NSMutableArray<QUIckControlValueTarget*> * targets;
@property (nonatomic, strong) NSMutableSet * scheduledActions;
@property (nonatomic, strong) NSMutableArray * actionTargets;
@end

@implementation QUIckControl

-(instancetype)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        [self loadStorages];
        [self registerExistedStates];
    }
    
    return self;
}

-(instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self loadStorages];
        [self registerExistedStates];
    }
    
    return self;
}

-(void)registerExistedStates {
    [self registerState:UIControlStateSelected forBoolKeyPath:keyPath(UIControl, selected) inverted:NO];
    [self registerState:UIControlStateHighlighted forBoolKeyPath:keyPath(UIControl, highlighted) inverted:NO];
    [self registerState:UIControlStateDisabled forBoolKeyPath:keyPath(UIControl, enabled) inverted:YES];
}

-(void)loadStorages {
    self.states = [NSMutableArray array];
    self.thisValueTarget = [[QUIckControlValueTarget alloc] initWithTarget:self];
    
    self.targets = [NSMutableArray array];
    
    self.scheduledActions = [NSMutableSet set];
    self.actionTargets = [NSMutableArray array];
}

#pragma mark - States

-(void)setSelected:(BOOL)selected {
    if (self.selected != selected) {
        [super setSelected:selected];
        [self applyCurrentState];
    }
}

-(void)setEnabled:(BOOL)enabled {
    if (self.enabled != enabled) {
        [super setEnabled:enabled];
        [self applyCurrentState];
    }
}

-(void)setHighlighted:(BOOL)highlighted {
    if (self.highlighted != highlighted) {
        [super setHighlighted:highlighted];
        [self applyCurrentState];
    }
}

-(void)beginTransition {
    self.isTransitionTime = YES;
}

-(void)commitTransition {
    if (!self.isTransitionTime) return;
    
    self.isTransitionTime = NO;
    [self applyCurrentState];
    for (NSNumber * action in self.scheduledActions) {
        [self sendActionsForControlEvents:[action unsignedIntegerValue]];
    }
    [self.scheduledActions removeAllObjects];
}

-(void)performTransition:(void(^)())transition {
    [self beginTransition];
    transition();
    [self commitTransition];
}

-(void)registerState:(UIControlState)state forBoolKeyPath:(NSString*)keyPath inverted:(BOOL)inverted {
    // & UIControlStateApplication ?
    QUICStateObject * stateObject = [QUICStateObject stateWithProperty:keyPath quickControlState:QUIckControlStateMake(state, inverted)];
    [self.states addObject:stateObject];
}

#pragma mark - Actions

-(QUIckControlActionTarget)addAction:(void (^)(__kindof QUIckControl *__weak))action forControlEvents:(UIControlEvents)events {
    QUIckControlActionTargetImp * actionTarget = [[QUIckControlActionTargetImp alloc] initWithControl:self controlEvents:events];
    actionTarget.action = action;
    [self.actionTargets addObject:actionTarget];
    
    return actionTarget;
}

-(QUIckControlActionTarget)addActionTarget:(QUIckControlActionTarget)target {
    QUIckControlActionTargetImp * sourceTarget = (QUIckControlActionTargetImp*)target;
    QUIckControlActionTargetImp * actionTarget = [[QUIckControlActionTargetImp alloc] initWithControl:self controlEvents:sourceTarget.events];
    actionTarget.action = sourceTarget.action;
    [self.actionTargets addObject:actionTarget];
    
    return actionTarget;
}

-(void)sendActionsForControlEvents:(UIControlEvents)controlEvents {
    if (self.isTransitionTime) { [self.scheduledActions addObject:@(controlEvents)]; return; }
    
    [super sendActionsForControlEvents:controlEvents];
}

#pragma mark - Values

-(void)removeValuesForTarget:(id)target {
    NSUInteger targetIndex = [self indexOfTarget:target];
    if (targetIndex != NSNotFound) {
        [self.targets removeObjectAtIndex:targetIndex];
    }
}

-(void)setValue:(id)value forTarget:(id)target forKeyPath:(NSString *)key forInvertedState:(UIControlState)state {
    [self setValue:value forTarget:target forKeyPath:key forStateDescriptor:QUICStateDescriptorMake(state, QUICStateTypeInverted)];
}

-(void)setValue:(id)value forTarget:(id)target forKeyPath:(NSString *)key forAllStatesContained:(UIControlState)state {
    [self setValue:value forTarget:target forKeyPath:key forStateDescriptor:QUICStateDescriptorMake(state, QUICStateTypeIntersected)];
}

-(void)setValue:(id)value forKeyPath:(NSString *)key forState:(UIControlState)state {
    [self setValue:value forTarget:self forKeyPath:key forState:state];
}

// TODO: Create possible add values for multiple states
-(void)setValue:(id)value forTarget:(id)target forKeyPath:(NSString *)key forState:(UIControlState)state {
    [self setValue:value forTarget:target forKeyPath:key forStateDescriptor:QUICStateDescriptorMake(state, QUICStateTypeUsual)];
}

-(void)setValue:(id)value forTarget:(id)target forKeyPath:(NSString *)key forStateDescriptor:(QUICStateDescriptor)descr {
    QUIckControlValueTarget * valueTarget = [self valueTargetForTarget:target];
    [valueTarget setValue:value forKeyPath:key forStateDescriptor:descr];
    if (QUICStateEvaluateWithState(descr, self.state)) {
        [valueTarget applyValue:value forKey:key];
    }
}

-(NSUInteger)indexOfTarget:(id)target {
    return [self.targets indexOfObjectPassingTest:^BOOL(QUIckControlValueTarget *  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if ([obj.target isEqual:target]) {
            *stop = YES;
            return YES;
        }
        return NO;
    }];
}

-(QUIckControlValueTarget*)valueTargetForTarget:(id)target {
    if (self == target) return self.thisValueTarget;
    
    NSUInteger index = [self indexOfTarget:target];
    if (index == NSNotFound) {
        QUIckControlValueTarget * valueTarget = [[QUIckControlValueTarget alloc] initWithTarget:target];
        [self.targets addObject:valueTarget];
        index = self.targets.count - 1;
    }
    
    return [self.targets objectAtIndex:index];
}

#pragma mark - Apply state values

-(void)applyCurrentStateForTarget:(id)target {
    [[self valueTargetForTarget:target] applyValuesForState:self.state];
}

-(void)applyState:(UIControlState)state {
    if (self.isTransitionTime) return;
    
    [self setNeedsDisplay];
    [self setNeedsLayout];
    
    [self.thisValueTarget applyValuesForState:state];
    for (QUIckControlValueTarget * target in self.targets) {
        [target applyValuesForState:state];
    }
}

-(id)valueForTarget:(id)target forKey:(NSString*)key forState:(UIControlState)state {
    return [[self valueTargetForTarget:target] valueForKey:key forState:state];
}

-(void)applyCurrentState {
    [self applyState:self.state];
}

- (UIControlState)state {
    UIControlState state = UIControlStateNormal;
    
    for (QUICStateObject * stateValue in self.states) {
        if ([stateValue evaluateWithObject:self]) {
            state |= stateValue.state.controlState;
        }
    }
    
    return state;
}

-(void)dealloc {
    for (QUIckControlActionTarget target in self.actionTargets) {
        [target stop];
    }
    [self.actionTargets removeAllObjects];
}

@end

File: QUIckControl/QUIckControl.swift
//
//  QUIckControl.swift
//  QUIckControl
//
//  Created by Denis Koryttsev on 23/10/16.
//  Copyright © 2016 Denis Koryttsev. All rights reserved.
//

import UIKit
import Statable

public protocol QUIckControlActionTarget {
    func start()
    func stop()
}

open class QUIckControl : UIControl, KnownStatable {
    public typealias Factor = QUIckControlStateFactor<QUIckControl>
    public typealias StateType = UIControlState
    
    public var isTransitionTime = false
    private lazy var thisValueTarget: QUIckControlValueTarget = QUIckControlValueTarget(target: self)
    
    public var factors = [Factor]()
    private var targets = [QUIckControlValueTarget]()
    private let scheduledActions = NSMutableSet()
    private var actionTargets = [QUIckControlActionTarget]()
    private lazy var subscribers: [QUIckControlSubscriber] = [QUIckControlSubscriber]()
    public var lastAppliedState: UIControlState = .normal {
        didSet {
            for subscriber in subscribers {
                subscriber.invoke(ifMatched: lastAppliedState)
            }
        }
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        
        registerExistedStates()
    }
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        
        registerExistedStates()
    }
    
    private func registerExistedStates() {
        register(.selected, forBoolKeyPath: #keyPath(UIControl.selected), inverted: false)
        register(.highlighted, forBoolKeyPath: #keyPath(UIControl.highlighted), inverted: false)
        register(.disabled, forBoolKeyPath: #keyPath(UIControl.enabled), inverted: true)
    }
    
    // MARK: States
    
    override open var isSelected: Bool {
        didSet { if oldValue != isSelected { applyCurrentState() } }
    }

    override open var isEnabled: Bool {
        didSet { if oldValue != isEnabled { applyCurrentState() } }
    }
    
    override open var isHighlighted: Bool {
        didSet { if oldValue != isHighlighted { applyCurrentState() } }
    }
    
    public func beginTransition() {
        isTransitionTime = true
    }
    
    public func endTransition() {
        isTransitionTime = false
        scheduledActions.removeAllObjects()
    }
    
    public func commitTransition() {
        if !isTransitionTime { return }
        
        isTransitionTime = false
        applyCurrentState()
        for action: Any in scheduledActions {
            sendActions(for: action as! UIControlEvents)
        }
        scheduledActions.removeAllObjects()
    }
    
    public func performTransition(withCommit commit: Bool = true, transition: () -> Void) {
        beginTransition()
        transition()
        commit ? commitTransition() : endTransition()
    }
    
    public func register(_ state: UIControlState, forBoolKeyPath keyPath: String, inverted: Bool) {
        // & UIControlStateApplication ?
//        factors.append(QBoolStateFactor(property: keyPath, state: state, inverted: inverted))
        factors.append(QUIckControlStateFactor(state: state, predicate: NSPredicate(format: "\(keyPath) == \(inverted ? "NO" : "YES")")))
        
        // example use block factor
//        let boolFactor = QBoolStateFactor(property: keyPath, state: state, inverted: inverted)
//        factors.append(StateFactor(state: state, predicate: { (control) -> Bool in
//            return boolFactor.evaluate(with: control)
//        }))
    }
    
    public func register(_ state: UIControlState, with predicate: NSPredicate) {
        factors.append(QUIckControlStateFactor(state: state, predicate: predicate))
    }
    
    // example use block factor
//    func register(_ state: UIControlState, with factor: @escaping (_ object: QUIckControl) -> Bool) {
//        factors.append(StateFactor(state: state, predicate: factor))
//    }
    
    // MARK: - Actions
    
    open func subscribe(on events: UIControlEvents, _ action: @escaping (QUIckControl) -> Void) -> QUIckControlActionTarget {
        let actionTarget = QUIckControlActionTargetImp(control: self, controlEvents: events)
        actionTarget.action = action
        actionTargets.append(actionTarget)
        return actionTarget
    }
    
    override open func sendActions(for controlEvents: UIControlEvents) {
        if isTransitionTime {
            scheduledActions.add(controlEvents)
            return
        }
        super.sendActions(for: controlEvents)
    }
    
    open func subscribe(on state: QUICStateDescriptor, _ action: @escaping () -> ()) {
        self.subscribers.append(QUIckControlSubscriber(for: state, action: action))
    }
    
    // MARK: - Values
    
    public func removeValues(forTarget target: NSObject, forKeyPath key: String, forState state: UIControlState) {
        valueTarget(forTarget: target).removeValues(for: key, forState: state)
    }
    
    public func removeValues(forTarget target: NSObject, forKeyPath key: String) {
        valueTarget(forTarget: target).removeValues(for: key)
    }
    
    public func removeValues(forTarget target: NSObject? = nil) {
        guard let externalTarget = target else { thisValueTarget.removeValues(); return }
        
        let targetIndex = indexOfTarget(externalTarget)
        if targetIndex != nil {
            targets.remove(at: targetIndex!)
        }
    }
    
    public func setValue(_ value: Any?, forTarget target: NSObject, forKeyPath key: String, forInvertedState state: UIControlState) {
        setValue(value, forTarget: target, forKeyPath: key, for: QUICStateDescriptor(inverted: state))
    }
    
    public func setValue(_ value: Any?, forTarget target: NSObject, forKeyPath key: String, forAllStatesContained state: UIControlState) {
        setValue(value, forTarget: target, forKeyPath: key, for: QUICStateDescriptor(intersected: state))
    }
    
    public func setValue(_ value: Any?, forKeyPath key: String, for state: UIControlState) {
        setValue(value, forTarget: self, forKeyPath: key, for: state)
    }
    
    public func setValue(_ value: Any?, forTarget target: NSObject, forKeyPath key: String, for state: UIControlState) {
        setValue(value, forTarget: target, forKeyPath: key, for: QUICStateDescriptor(usual: state))
    }
    
    public func setValue(_ value: Any?, forTarget target: NSObject? = nil, forKeyPath key: String, for descriptor: QUICStateDescriptor) {
        let valTarget = valueTarget(forTarget: target ?? self)
        valTarget.setValue(value, forKeyPath: key, for: descriptor)
        if descriptor.evaluate(with: state) {
            valTarget.applyValue(value, forKey: key)
        }
    }
    
    private func indexOfTarget(_ target: NSObject) -> Int? {
        return targets.index(where: { return $0.target.isEqual(target) })
    }
    
    private func valueTarget(forTarget target: NSObject!) -> QUIckControlValueTarget {
        if self == target { return thisValueTarget }
        
        var index = indexOfTarget(target)
        if index == nil {
            let valueTarget = QUIckControlValueTarget(target: target)
            targets.append(valueTarget)
            index = targets.count - 1
        }
        return targets[index!]
    }
    
    // MARK: - Apply state values
    
    open func applyCurrentState() {
        guard !isTransitionTime else { return }
        
        let currentState = state
        guard lastAppliedState != currentState else { return }
        
        apply(state: currentState)
        lastAppliedState = currentState
    }
    
    open func applyCurrentState(forTarget target: NSObject) {
        valueTarget(forTarget: target).apply(state: state)
    }
    
    // force apply state
    open func apply(state: UIControlState) {
        setNeedsDisplay()
        setNeedsLayout()
        thisValueTarget.apply(state: state)
        for target: QUIckControlValueTarget in targets {
            target.apply(state: state)
        }
    }
    
    public func value(for target: NSObject, forKey key: String, for state: UIControlState) -> Any? {
        return valueTarget(forTarget: target).valueForKey(key: key, forState: state)
    }
    
    override open var state: UIControlState {
        var result: UIControlState = .normal
        for factor in factors {
            factor.mark(state: &result, ifEvaluatedWith: self)
        }
        return result
    }

    // MARK: Deinitialier
    
    deinit {
        for target in actionTargets {
            target.stop()
        }
        actionTargets.removeAll()
    }
}
